bsmetadata/metadata_utils.py

```diff
@@ -122,6 +142,71 @@ def create_global_metadata_prefix(example: Dict[str, Any], cfg: MetadataConfig)
     return cfg.metadata_sep.join(sorted_metadata) + cfg.global_metadata_sep if sorted_metadata else ""
 
 
+def _collate_metadata(metadata_list: List[dict], cfg: MetadataConfig):
+    """Transforms a list of local metadata that may be more than one at a same `char_start_idx` or `char_end_idx` index
+    into a list of metadata with only one metadata per `char_start_idx`.
+    The new metadata at `char_start_idx` corresponds to the concatenation of the local metadata that appear at the
+    same idx (for its start value and/or its end value). The order is determined by the values of `relative_start_pos` 
+    and `relative_end_pos`.
+
+    Note that this function requires that for each metadata have `char_start_idx`, `char_end_idx`,
+    `relative_start_pos` and `relative_end_pos`
+
+    Args:
+        metadata_list: [description]
+        cfg: The data config to use.
+
+    Returns:
+        A list of metadata with only one metadata per `char_start_idx`
+    """
+    processor = PROCESSORS.get(metadata_list[0]["key"], MetadataProcessor)(cfg)
+
+    new_metadata_list = []
+
+    metadata_dict_idx = DefaultDict(dict)
+    for metadata_node in metadata_list:
+        processed_metadata = processor.process_local(metadata_node)
+        if processed_metadata is None:
+            continue
+
+        metadata_node = BasicMetadata(
+            char_start_idx=metadata_node["char_start_idx"],
+            key=metadata_node["key"],
+            type=metadata_node["type"],
+            value=metadata_node["value"],
+            char_end_idx=metadata_node["char_end_idx"],
+            relative_start_pos=metadata_node["relative_start_pos"],
+            relative_end_pos=metadata_node["relative_end_pos"],
+        )
```

Comment:
Personally, I'd prefer a solution where *everything* that is metadata-specific is stored in the ``value`` field.
For example, instead of

```json
{
    "key":"html",
    "type":"local",
    "value":"i",
    "html_attrs":{
        "attrs":[
            "class"
        ],
        "values":[
            "level2"
        ]
    }
}
```
one could write
```json
{
    "key":"html",
    "type":"local",
    "value":{
        "tag":"i",
        "attrs":[
            "class"
        ],
        "values":[
            "level2"
        ]
    }
}
```
This would feel cleaner to me (and, as a side effect, it would allow the solution I've mentioned above). What do you think about that?

Reply:
I had actually also started with the last format you propose. But I changed it to the current one with `html_attrs` because datasets needs to have a unique type for all values in a column. In other words, dataset does not accept strings for some metadata types and dictionaries for others.

I also think it would be better if everything was in the `value` field, but this would require a format change for other types of metadata.

Reply:
Ah, I see - I wasn't aware of this limitation of `datasets`. This might also be relevant for other metadata types where the value isn't simply plain text (e.g., for entities we might have both a name and a description, and for data source we might have some hierarchical structure). 

It's probably easiest then to go with each metadata having its own `<metadata_key>_attrs` (`html_attrs`, `entity_attrs`, ...) as needed.